### ISSUE TITLE
Allow installed ruby option instead of packaged ruby

### ISSUE THREAD

**Issue by JeanmarcLaiAqovia** [SEEK-Jobs/pact-net#97]
Would it be possible to use a installed ruby version instead of packaging the ruby for each instance? the lib folder that gets packaged into the nuget package in folder build\core\pact\lib is 28.3 MB. We have a micro-services architecture so would rather use one installed ruby instead as we have about >30 micro-services and growing.

**Comment by neilcampbell** [comment]
The actual .nupkg with the Ruby core is about 10MB (because it's zipped), but yeah as you mention when it's unzipped it is about 30MB.

I'm interested to understand why the 30MB is a concern for you? Are you committing the nuget packages to your source control?

Basically my thoughts were that anywhere the core is copied shouldn't need to be checked into source control, so size shouldn't really be a major concern. On top of that it's a testing library so Pact-Net should only be referenced by test projects and as such it should never get shipped to a prod environment or be bundled into any production build artefacts.

Interested to hear more.

**Comment by ma499** [comment]
I work with @JeanmarcLaiAqovia. Here are my thoughts although none of them are showstoppers for us at the moment:

- I haven't checked if the Pact.NET library is .NET Standard yet but eventually I hope it is and we'll need it to be cross-platform to be more broadly useful with ASP.NET Core which I suspect most .NET microservices teams will be evolving towards. 
- currently the 30 MB per service would gives us about 2 GB of wasted space (and growing) on the whole stack. It seems a bit wasteful but is not the end of the world.

We don't commit nuget packages but I'm aware of others that do :-)

If we can agree that this is a good idea, and an approach to implementing it, would you be happy to receive a pull request? To my mind we could either pass in a ruby path as a config parameter or expect an environment variable to be set. Perhaps we could have two flavours of the nuget package since I realise many .NET teams won't want to mess around with a ruby install.

**Comment by neilcampbell** [comment]
@ma499 Thanks for the additional context.
I am completely ok with doing this if we can come up with a way to support both, and yes more than happy to accept PRs 😄

To confirm are you wanting to install Ruby or install the Pact Ruby Standalone core entirely in a shared location?

I would love to start targeting .NET standard soon'ish, and also think about cross platform support. For the Ruby standalone core to work cross platform, it requires a separate package for Linux, OSX and Windows and if we were to keep using the current convention it would further exacerbate the problem. Another idea I had was to download the Ruby shared core on demand (maybe in a nuget init script, apparently install scripts are not longer supported in Nuget 3.x) depending on the platform it is currently running on, but that also has trade offs like requiring internet access (if a company uses an internal nuget cache). Any other ideas you have on this?

I think if we can come up with a strategy for the above, we can then come up with a strategy for running ruby in a non-standalone way fairly easily.

**Comment by adamrodger** [comment]
I think the best idea is to generate a 'plain' version which relies on the user having Ruby and Pact installed locally and on the PATH, and 'standalone' NuGets for each OS which package the appropriate Ruby Core as `contentFiles` - i.e. a `PactNet.nupkg` which just has `PactNet.dll` and a `PactNet.Win32`/`PactNet.Linux` etc which just have Ruby Core in them.

PactNet would then first look in the local directory to see if it can find a standalone version, and if it can't then it tries to use the version on PATH. If that doesn't work then you just throw up an error saying you either need to install one of the OS-specific packages or install Ruby/Pact manually.

**Comment by adamrodger** [comment]
I was thinking a NuSpec file something like this with a version for each OS:
```xml
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2012/06/nuspec.xsd">
  <metadata>
    <id>Pact.Standalone.Win32</id>
    <version>$version$</version>
    <authors>Pact Foundation</authors>
    <owners>Pact Foundation</owners>
    <requireLicenseAcceptance>false</requireLicenseAcceptance>
    <licenseUrl>https://github.com/pact-foundation/pact-ruby-standalone/blob/master/LICENSE</licenseUrl>
    <projectUrl>https://github.com/pact-foundation/pact-ruby-standalone</projectUrl>
    <description>Win32 version of Ruby Pact Core</description>
    <tags>pact</tags>
    <repository url="https://github.com/pact-foundation/pact-ruby-standalone" />
    <contentFiles>
      <files include="**" buildAction="None" copyToOutput="true" flatten="false" />
    </contentFiles>
  </metadata>
  <files>
    <file src="standalone\pact\**\*.*" target="contentFiles\any\any\" />
  </files>
</package>
```
PactNet could then check if one of those is installed by looking for a `.\pact` folder and if it doesn't find one then try and use one from the PATH. That should give you the best of both worlds if people want to use the standalone version but without making the NuGet package huge for people that don't.

You'd have to get permission from the upstream Pact devs though because technically it's them that should redistribute the standalone version, not PactNet.

Note: I've not tested that this actually works, but it should in theory.

**Comment by neilcampbell** [comment]
@adamrodger Yeah that was another idea I was tossing around in my head. I think that is quite neat, and should work fine as long as we nail the error messages and doco to instruct the user what to do.

Sound like we will need to extract the actual PactNet ruby code out of the current zip and package into the PactNet library, that way we can support running with an installed version of Ruby and no other package.

One downside I can think of is what happens when someone develops on Windows or OSX and deploys to Linux?

**Comment by neilcampbell** [comment]
@bethesque It would be awesome to get your take on this and if you have any other ideas?

**Comment by timrog1** [comment]
I was talking with @ma499 and @JeanmarcLaiAqovia about this, and @adamrodger 's solution sounds the best idea. I wondered if there was any reason you thought the standalone version should take precedence over the installed one? Other than it all being nicely self-contained, are there any other reasons to prefer a standalone version? We'd see it working the other way round: the installed one is the one that would be configured with server certificates etc, so we'd prefer that to take precedence.

@neilcampbell I guess those hoping to run on both Windows and Linux could install both Windows and Linux packages and it picks the appropriate one at runtime in the absence of an installed version? Although, I'd imagine Linux users were much more likely to have Ruby installed.

**Comment by adamrodger** [comment]
@timrog1 The reason I think the standalone version one should be favoured is that if someone has gone to the effort of including the 'native' standalone version, that means they really must want to use it. It wouldn't make sense to add that as a NuGet reference only to ignore it and use a native one instead.

If you've configured a native one with SSL certs and things, you wouldn't install the standalone version.

**Comment by bethesque** [comment]
Not knowing how NuGet works, I can explain how pact-js does it. There is a separate pact-js package for each OS, as well as a plain pact-js package. When pact-js is installed, it detects the operating system on install, then picks the appropriate child package to install. I don't know if this kind of thing would work with NuGet though.

**Comment by neilcampbell** [comment]
@bethesque Thanks for that info. Does the plain pact-js work without any of the OS specific packages and use the locally install Ruby? If so how does it include the ruby code?

Unfortunately Nuget is deprecating the ability to run scripts on install, so likely we will need to document that devs will need to install the base package (PactNet) and then the OS specific standalone (which probably wont include the PactNet code itself, just the travelling Ruby bits) or install ruby.

**Comment by bethesque** [comment]
There is no option to use a local ruby installation in pact-js.

Could you reverse the dependency and have "pact.net for osx", "pact.net for windows", "pact.net standalone" etc which then installed the common package? That way you don't have to install two things, you just pick the appropriate one for your OS.

**Comment by neilcampbell** [comment]
Yeah we can definitely do something like that as well. We just need to make sure we handle the scenario where you Dev on OSX and deploy to Linux (or any other combination), so allowing multiple language specific packages to be install at the same time.

I would imagine something like below:
- PactNet
  Contains the core PactNet code and allows you to run on a locally Ruby install and also discover any installed OS specific package (based on current OS) and run without a locally install Ruby.
- PactNet-OSX (depends on PactNet)
  Contains travelling ruby for OSX and any bindings
- PactNet-Win (depends on PactNet)
  Contains travelling ruby for Windows and any bindings
- PactNet-Linux (depends on PactNet)
  Contains travelling ruby for Linux and any bindings

This also handles the scenario when someone is updating, we would simply throw an exception if there is no Ruby installed locally and no OS specific package with info about what to do. But yeah the doco would say to install PactNet-Win (or whatever), which would auto install the PactNet package.

**Comment by bethesque** [comment]
Sounds good. One extra implementation detail is that there are actually two linux packages - linux 32 bit and linux 64 bit.

**Comment by neilcampbell** [comment]
Now that we have .net core/standard support, who is keen to have a crack at working on this?

**Comment by neilcampbell** [comment]
Cross platform support is now merged into master (and released in the latest nuget) and we now have 3 new platform specific packages, which work as described above:
- PactNet-Windows
- PactNet-OSX
- PactNet-Linux-x64
- PactNet-Linux-x86

@JeanmarcLaiAqovia @ma499 If you would still like to use a local ruby install, I am happy to accept pull requests for this feature. There is a TODO placeholder for where this code would go see https://github.com/SEEK-Jobs/pact-net/blob/master/PactNet/Core/PactCoreHost.cs#L61

**Comment by adamrodger** [comment]
I know this is resurrecting an old issue, but now that we've got the mock service [builds as a Docker image](https://hub.docker.com/r/pactfoundation/pact-mock-service), I'm going to revisit creating some kind of `PactNet.Standalone` whereby you just point it at an existing running mock server and it no longer manages the lifetime of it.

For my uses, I'd like a `docker-compose` file or similar which starts the mock service in one container and runs my consumer tests in another container against that external service.

We'll also need the equivalent for provider tests of course, but there's also a [provider verifier image](https://hub.docker.com/r/dius/pact-provider-verifier-docker) as well so it could work in a similar way.

The overall aim has to be that we don't bundle the entire Ruby standalone version (since that makes the NuGets really big and not cross-platform) and give the possibility of delegating that problem to the user if they so wish. If they don't, the bundled versions are still available and PactNet will manage the lifecycle of the mock server/verifier processes for you.

**Comment by neilcampbell** [comment]
Cool, I like the idea!

A couple of questions/statements:
- Is size that is actually the problem that needs to be solved? 
- The ruby standalone should work fine already cross platform

To me the main problem we are actually trying to solve is that Windows has path limits and the ruby standalone has deeply nested path. Don't get me wrong though, It would be great if you only need to install what you use.

I don't think we want or need a seperate package, the core PactNet is bundled into a small seperate package already. At the moment though you need to install a seperate package (PactNet.Windows, PactNet.Linux, PactNet.OSX), which contains the ruby standalone bits and it has a dependency on the PactNet package. The way I was thinking, was we'd add a way to configure Pact to signal that you going to supply your own consumer mock and provider verifier, and we will just use that.

Would be awesome if you'd like to take a crack at it!

**Comment by adamrodger** [comment]
Yeah, from what I remember of that part of the code it looks for one of the `PactNet.$ARCH` sub-packages and then bombs out if it can't find one. What I'm thinking is that we just need to be able to say to that part "look, I know there's no arch-specific package installed, and that's fine - use this server over here that I've already started for you"

That way the 'external' mock server could either be running the Docker container, or you could've started it manually from an existing Ruby installation - anything really. The point is more that we delegate that responsibility to the user if they're willing to accept it.

Re: package size - yes they're a bit of a problem for us (although not a huge one):

- Our CI pipeline caches artifacts between stages, and transferring and/or zipping all those tiny files through the packaged Ruby runtime makes it really slow
- If they're baked into a Docker image, it makes the image really big also. We're using multi-stage builds to get around that currently so they're not added to the final image.

I'll let you know how I get on anyway 👍 Hopefully there'll be a PR to follow.

**Comment by neilcampbell** [comment]
@adamrodger Yeah was thinking exactly the same thing. Shout if you have any questions.

**Comment by vpsouza** [comment]
Devs on my team are using Windows, Mac and Linux OSs. And our CI pipeline runs on Linux. How can we proceed to run locally and on CI no different OSs?

**Comment by bethesque** [comment]
The above plan was never implemented, as far as I know. It's a single package that works on all platforms.
